# OneHot Encoding

# Dummy Variables and OneHot Encoding

In [54]:
#Importing essential libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [55]:
#Loading Dataset
#df=sns.load_dataset('tips')
tips_df=pd.read_csv("tips.csv")
tips_df

Unnamed: 0  total_bill    tip     sex smoker   day    time  size
0             0       16.99   1.01  Female     No   Sun  Dinner     2
1             1       10.34   1.66    Male     No   Sun  Dinner     3
2             2       21.01   3.50    Male     No   Sun  Dinner     3
3             3       23.68   3.31    Male     No   Sun  Dinner     2
4             4       24.59   3.61  Female     No   Sun  Dinner     4
5             5       25.29   4.71    Male     No   Sun  Dinner     4
6             6        8.77   2.00    Male     No   Sun  Dinner     2
7             7       26.88   3.12    Male     No   Sun  Dinner     4
8             8       15.04   1.96    Male     No   Sun  Dinner     2
9             9       14.78   3.23    Male     No   Sun  Dinner     2
10           10       10.27   1.71    Male     No   Sun  Dinner     2
11           11       35.26   5.00  Female     No   Sun  Dinner     4
12           12       15.42   1.57    Male     No   Sun  Dinner     2
13           13       18.43   3.00    Male     No   Sun  Dinner     4
14           14       14.83   3.02  Female     No   Sun  Dinner     2
15           15       21.58   3.92    Male     No   Sun  Dinner     2
16           16       10.33   1.67  Female     No   Sun  Dinner     3
17           17       16.29   3.71    Male     No   Sun  Dinner     3
18           18       16.97   3.50  Female     No   Sun  Dinner     3
19           19       20.65   3.35    Male     No   Sat  Dinner     3
20           20       17.92   4.08    Male     No   Sat  Dinner     2
21           21       20.29   2.75  Female     No   Sat  Dinner     2
22           22       15.77   2.23  Female     No   Sat  Dinner     2
23           23       39.42   7.58    Male     No   Sat  Dinner     4
24           24       19.82   3.18    Male     No   Sat  Dinner     2
25           25       17.81   2.34    Male     No   Sat  Dinner     4
26           26       13.37   2.00    Male     No   Sat  Dinner     2
27           27       12.69   2.00    Male     No   Sat  Dinner     2
28           28       21.70   4.30    Male     No   Sat  Dinner     2
29           29       19.65   3.00  Female     No   Sat  Dinner     2
30           30        9.55   1.45    Male     No   Sat  Dinner     2
31           31       18.35   2.50    Male     No   Sat  Dinner     4
32           32       15.06   3.00  Female     No   Sat  Dinner     2
33           33       20.69   2.45  Female     No   Sat  Dinner     4
34           34       17.78   3.27    Male     No   Sat  Dinner     2
35           35       24.06   3.60    Male     No   Sat  Dinner     3
36           36       16.31   2.00    Male     No   Sat  Dinner     3
37           37       16.93   3.07  Female     No   Sat  Dinner     3
38           38       18.69   2.31    Male     No   Sat  Dinner     3
39           39       31.27   5.00    Male     No   Sat  Dinner     3
40           40       16.04   2.24    Male     No   Sat  Dinner     3
41           41       17.46   2.54    Male     No   Sun  Dinner     2
42           42       13.94   3.06    Male     No   Sun  Dinner     2
43           43        9.68   1.32    Male     No   Sun  Dinner     2
44           44       30.40   5.60    Male     No   Sun  Dinner     4
45           45       18.29   3.00    Male     No   Sun  Dinner     2
46           46       22.23   5.00    Male     No   Sun  Dinner     2
47           47       32.40   6.00    Male     No   Sun  Dinner     4
48           48       28.55   2.05    Male     No   Sun  Dinner     3
49           49       18.04   3.00    Male     No   Sun  Dinner     2
50           50       12.54   2.50    Male     No   Sun  Dinner     2
51           51       10.29   2.60  Female     No   Sun  Dinner     2
52           52       34.81   5.20  Female     No   Sun  Dinner     4
53           53        9.94   1.56    Male     No   Sun  Dinner     2
54           54       25.56   4.34    Male     No   Sun  Dinner     4
55           55       19.49   3.51    Male     No   Sun  Dinner     2
56           56

In [56]:
#pd.set_option("display.max_columns",None)
#pd.set_option("display.max_rows",None)
#df

In [57]:
df.head()

Unnamed: 0  total_bill   tip     sex smoker  day    time  size
0           0       16.99  1.01  Female     No  Sun  Dinner     2
1           1       10.34  1.66    Male     No  Sun  Dinner     3
2           2       21.01  3.50    Male     No  Sun  Dinner     3
3           3       23.68  3.31    Male     No  Sun  Dinner     2
4           4       24.59  3.61  Female     No  Sun  Dinner     4

In [58]:
df.tail()

Unnamed: 0  total_bill   tip     sex smoker   day    time  size
239         239       29.03  5.92    Male     No   Sat  Dinner     3
240         240       27.18  2.00  Female    Yes   Sat  Dinner     2
241         241       22.67  2.00    Male    Yes   Sat  Dinner     2
242         242       17.82  1.75    Male     No   Sat  Dinner     2
243         243       18.78  3.00  Female     No  Thur  Dinner     2

In [59]:
tips_df.shape

(244, 8)

# Now we have data in which we have 244 rows and 8 columns are present.

Now in this we have to convert our categorical data into dummy data.

In [64]:
dummy_df=pd.get_dummies(tips_df)
dummy_df

Unnamed: 0  total_bill    tip  size  sex_Female  sex_Male  smoker_No  \
0             0       16.99   1.01     2           1         0          1   
1             1       10.34   1.66     3           0         1          1   
2             2       21.01   3.50     3           0         1          1   
3             3       23.68   3.31     2           0         1          1   
4             4       24.59   3.61     4           1         0          1   
5             5       25.29   4.71     4           0         1          1   
6             6        8.77   2.00     2           0         1          1   
7             7       26.88   3.12     4           0         1          1   
8             8       15.04   1.96     2           0         1          1   
9             9       14.78   3.23     2           0         1          1   
10           10       10.27   1.71     2           0         1          1   
11           11       35.26   5.00     4           1         0          1   
12           12       15.42   1.57     2           0         1          1   
13           13       18.43   3.00     4           0         1          1   
14           14       14.83   3.02     2           1         0          1   
15           15       21.58   3.92     2           0         1          1   
16           16       10.33   1.67     3           1         0          1   
17           17       16.29   3.71     3           0         1          1   
18           18       16.97   3.50     3           1         0          1   
19           19       20.65   3.35     3           0         1          1   
20           20       17.92   4.08     2           0         1          1   
21           21       20.29   2.75     2           1         0          1   
22           22       15.77   2.23     2           1         0          1   
23           23       39.42   7.58     4           0         1          1   
24           24       19.82   3.18     2           0         1          1   
25           25       17.81   2.34     4           0         1          1   
26           26       13.37   2.00     2           0         1          1   
27           27       12.69   2.00     2           0         1          1   
28           28       21.70   4.30     2           0         1          1   
29           29       19.65   3.00     2           1         0          1   
30           30        9.55   1.45     2           0         1          1   
31           31       18.35   2.50     4           0         1          1   
32           32       15.06   3.00     2           1         0          1   
33           33       20.69   2.45     4           1         0          1   
34           34       17.78   3.27     2           0         1          1   
35           35       24.06   3.60     3           0         1          1   
36           36       16.31   2.00     3           0         1          1   
37           37       16.93   3.07     3           1         0          1   
38           38       18.69   2.31     3           0         1          1   
39           39       31.27   5.00     3           0         1          1   
40           40       16.04   2.24     3           0         1          1   
41           41       17.46   2.54     2           0         1          1   
42           42       13.94   3.06     2           0         1          1   
43           43        9.68   1.32     2           0         1          1   
44           44       30.40   5.60     4           0         1          1   
45           45       18.29   3.00     2           0         1          1   
46           46       22.23   5.00     2           0         1          1   
47           47       32.40   6.00     4           0         1          1   
48           48       28.55   2.05     3           0         1          1   
49           49       18.04   3.00     2           0         1          1   
50           50       12.54   2.50     2           0         1          1   
5

In [65]:
dummy_df.shape

(244, 14)

In [66]:
pd.get_dummies(tips_df,drop_first=True)
#To reduce features with common properties for to reduce overload from model

Unnamed: 0  total_bill    tip  size  sex_Male  smoker_Yes  day_Sat  \
0             0       16.99   1.01     2         0           0        0   
1             1       10.34   1.66     3         1           0        0   
2             2       21.01   3.50     3         1           0        0   
3             3       23.68   3.31     2         1           0        0   
4             4       24.59   3.61     4         0           0        0   
5             5       25.29   4.71     4         1           0        0   
6             6        8.77   2.00     2         1           0        0   
7             7       26.88   3.12     4         1           0        0   
8             8       15.04   1.96     2         1           0        0   
9             9       14.78   3.23     2         1           0        0   
10           10       10.27   1.71     2         1           0        0   
11           11       35.26   5.00     4         0           0        0   
12           12       15.42   1.57     2         1           0        0   
13           13       18.43   3.00     4         1           0        0   
14           14       14.83   3.02     2         0           0        0   
15           15       21.58   3.92     2         1           0        0   
16           16       10.33   1.67     3         0           0        0   
17           17       16.29   3.71     3         1           0        0   
18           18       16.97   3.50     3         0           0        0   
19           19       20.65   3.35     3         1           0        1   
20           20       17.92   4.08     2         1           0        1   
21           21       20.29   2.75     2         0           0        1   
22           22       15.77   2.23     2         0           0        1   
23           23       39.42   7.58     4         1           0        1   
24           24       19.82   3.18     2         1           0        1   
25           25       17.81   2.34     4         1           0        1   
26           26       13.37   2.00     2         1           0        1   
27           27       12.69   2.00     2         1           0        1   
28           28       21.70   4.30     2         1           0        1   
29           29       19.65   3.00     2         0           0        1   
30           30        9.55   1.45     2         1           0        1   
31           31       18.35   2.50     4         1           0        1   
32           32       15.06   3.00     2         0           0        1   
33           33       20.69   2.45     4         0           0        1   
34           34       17.78   3.27     2         1           0        1   
35           35       24.06   3.60     3         1           0        1   
36           36       16.31   2.00     3         1           0        1   
37           37       16.93   3.07     3         0           0        1   
38           38       18.69   2.31     3         1           0        1   
39           39       31.27   5.00     3         1           0        1   
40           40       16.04   2.24     3         1           0        1   
41           41       17.46   2.54     2         1           0        0   
42           42       13.94   3.06     2         1           0        0   
43           43        9.68   1.32     2         1           0        0   
44           44       30.40   5.60     4         1           0        0   
45           45       18.29   3.00     2         1           0        0   
46           46       22.23   5.00     2         1           0        0   
47           47       32.40   6.00     4         1           0        0   
48           48       28.55   2.05     3         1           0        0   
49           49       18.04   3.00     2         1           0        0   
50           50       12.54   2.50     2         1           0        0   
51           51       10.29   2.60     2         0           0        0   
52           52       34.81   

# Now we will convert it into one hot encoder

In [22]:
from sklearn.preprocessing import OneHotEncoder

In [67]:
oh_enc=OneHotEncoder(sparse=False)#Sparse:bool,default=True
#Return Matrix if sparse=True otherwise it will return an array when sparse=False
oh_enc

OneHotEncoder(sparse=False)

In [68]:
oh_enc_arr=oh_enc.fit_transform(tips_df[['sex','smoker','day','time']])
oh_enc_arr #Now we got data in numpy array and when we do get dummies we get data in array format

array([[1., 0., 1., ..., 0., 1., 0.],
       [0., 1., 1., ..., 0., 1., 0.],
       [0., 1., 1., ..., 0., 1., 0.],
       ...,
       [0., 1., 0., ..., 0., 1., 0.],
       [0., 1., 1., ..., 0., 1., 0.],
       [1., 0., 1., ..., 1., 1., 0.]])

In [69]:
dummy_df.keys()

Index(['Unnamed: 0', 'total_bill', 'tip', 'size', 'sex_Female', 'sex_Male',
       'smoker_No', 'smoker_Yes', 'day_Fri', 'day_Sat', 'day_Sun', 'day_Thur',
       'time_Dinner', 'time_Lunch'],
      dtype='object')

In [70]:
#We will create dataframe using categorical features
oh_enc_df=pd.DataFrame(oh_enc_arr,columns=['sex_Female', 'sex_Male',
       'smoker_No', 'smoker_Yes', 'day_Fri', 'day_Sat', 'day_Sun', 'day_Thur',
       'time_Dinner', 'time_Lunch'])
oh_enc_df

sex_Female  sex_Male  smoker_No  smoker_Yes  day_Fri  day_Sat  day_Sun  \
0           1.0       0.0        1.0         0.0      0.0      0.0      1.0   
1           0.0       1.0        1.0         0.0      0.0      0.0      1.0   
2           0.0       1.0        1.0         0.0      0.0      0.0      1.0   
3           0.0       1.0        1.0         0.0      0.0      0.0      1.0   
4           1.0       0.0        1.0         0.0      0.0      0.0      1.0   
5           0.0       1.0        1.0         0.0      0.0      0.0      1.0   
6           0.0       1.0        1.0         0.0      0.0      0.0      1.0   
7           0.0       1.0        1.0         0.0      0.0      0.0      1.0   
8           0.0       1.0        1.0         0.0      0.0      0.0      1.0   
9           0.0       1.0        1.0         0.0      0.0      0.0      1.0   
10          0.0       1.0        1.0         0.0      0.0      0.0      1.0   
11          1.0       0.0        1.0         0.0      0.0      0.0      1.0   
12          0.0       1.0        1.0         0.0      0.0      0.0      1.0   
13          0.0       1.0        1.0         0.0      0.0      0.0      1.0   
14          1.0       0.0        1.0         0.0      0.0      0.0      1.0   
15          0.0       1.0        1.0         0.0      0.0      0.0      1.0   
16          1.0       0.0        1.0         0.0      0.0      0.0      1.0   
17          0.0       1.0        1.0         0.0      0.0      0.0      1.0   
18          1.0       0.0        1.0         0.0      0.0      0.0      1.0   
19          0.0       1.0        1.0         0.0      0.0      1.0      0.0   
20          0.0       1.0        1.0         0.0      0.0      1.0      0.0   
21          1.0       0.0        1.0         0.0      0.0      1.0      0.0   
22          1.0       0.0        1.0         0.0      0.0      1.0      0.0   
23          0.0       1.0        1.0         0.0      0.0      1.0      0.0   
24          0.0       1.0        1.0         0.0      0.0      1.0      0.0   
25          0.0       1.0        1.0         0.0      0.0      1.0      0.0   
26          0.0       1.0        1.0         0.0      0.0      1.0      0.0   
27          0.0       1.0        1.0         0.0      0.0      1.0      0.0   
28          0.0       1.0        1.0         0.0      0.0      1.0      0.0   
29          1.0       0.0        1.0         0.0      0.0      1.0      0.0   
30          0.0       1.0        1.0         0.0      0.0      1.0      0.0   
31          0.0       1.0        1.0         0.0      0.0      1.0      0.0   
32          1.0       0.0        1.0         0.0      0.0      1.0      0.0   
33          1.0       0.0        1.0         0.0      0.0      1.0      0.0   
34          0.0       1.0        1.0         0.0      0.0      1.0      0.0   
35          0.0       1.0        1.0         0.0      0.0      1.0      0.0   
36          0.0       1.0        1.0         0.0      0.0      1.0      0.0   
37          1.0       0.0        1.0         0.0      0.0      1.0      0.0   
38          0.0       1.0        1.0         0.0      0.0      1.0      0.0   
39          0.0       1.0        1.0         0.0      0.0      1.0      0.0   
40          0.0       1.0        1.0         0.0      0.0      1.0      0.0   
41          0.0       1.0        1.0         0.0      0.0      0.0      1.0   
42          0.0       1.0        1.0         0.0      0.0      0.0      1.0   
43          0.0       1.0        1.0         0.0      0.0      0.0      1.0   
44          0.0       1.0        1.0         0.0      0.0      0.0      1.0   
45          0.0       1.0        1.0         0.0      0.0      0.0      1.0   
46          0.0       1.0        1.0         0.0      0.0      0.0      1.0   
47          0.0       1.0        1.0         0.0      0.0      0.0      1.0   
48          0.0       1.0        1.0         0.0      0.0      0.0      1.0   
49          0.0       1.0        1.0         0.0      0

# How to drop

In [77]:
oh_enc=OneHotEncoder(sparse=False,drop='first')
oh_enc

OneHotEncoder(drop='first', sparse=False)

In [78]:
oh_enc_arr=oh_enc.fit_transform(tips_df[['sex','smoker','day','time']])
oh_enc_arr

array([[0., 0., 0., 1., 0., 0.],
       [1., 0., 0., 1., 0., 0.],
       [1., 0., 0., 1., 0., 0.],
       ...,
       [1., 1., 1., 0., 0., 0.],
       [1., 0., 1., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0.]])

In [79]:
dummy_df.keys()

Index(['Unnamed: 0', 'total_bill', 'tip', 'size', 'sex_Female', 'sex_Male',
       'smoker_No', 'smoker_Yes', 'day_Fri', 'day_Sat', 'day_Sun', 'day_Thur',
       'time_Dinner', 'time_Lunch'],
      dtype='object')

What is One Hot Encoding?

As a machine can only understand numbers and cannot understand the text in the first place, this essentially becomes the case with Deep Learning & Machine Learning algorithms. One hot encoding can be defined as the essential process of converting the categorical data variables to be provided to machine and deep learning algorithms which in turn improve predictions as well as classification accuracy of a model. One Hot Encoding is a common way of preprocessing categorical features for machine learning models. This type of encoding creates a new binary feature for each possible category and assigns a value of 1 to the feature of each sample that corresponds to its original category. 


One hot encoding is a highly essential part of the feature engineering process in training for learning techniques. For example, we had our variables like colors and the labels were “red,” “green,” and “blue,” we could encode each of these labels as a three-element binary vector as Red: [1, 0, 0], Green: [0, 1, 0], Blue: [0, 0, 1]. The Categorical data while processing, must be converted to a numerical form. One-hot encoding is generally applied to the integer representation of the data. Here the integer encoded variable is removed and a new binary variable is added for each unique integer value. During the process, it takes a column that has categorical data, which has been label encoded and then splits the following column into multiple columns. The numbers are replaced by 1s and 0s randomly, depending on which column has what value. While the method is helpful for some ordinal situations, some input data does not have any ranking for category values, and this can lead to issues with predictions and poor performance. 

https://www.geeksforgeeks.org/ml-one-hot-encoding-of-datasets-in-python/ 